software/controller/lib/hal/spi_chain.tpp

```diff
@@ -0,0 +1,235 @@
+  }
+
+  // Send a reset command to all the slaves.
+  for(size_t byte=0 ; byte < sizeof(probe_buffer) ; byte++){
+    probe_buffer[byte] = reset_command;
+  }
+  SendDataWithBusyWait(probe_buffer, response_buffer, sizeof(probe_buffer));
+
+  // At this point, the response buffer should be filled with as many null commands as there are
+  // slaves.
+  num_slaves_ = 0;
+  for (uint8_t i : response_buffer) {
+    if (i == null_command){
+      num_slaves_++;
+    } else {
+      break;
+    }
+  }
+
+  // If the response buffer is full of null commands, this most likely means the
+  // daisy chain is broken (more slaves than we expect)
+  if(num_slaves_==sizeof(response_buffer)) {
+    num_slaves_ = 0;
+  }
+
+}
+
+template<size_t MaxSlaves>
+void DaisyChain<MaxSlaves>::SendDataWithBusyWait(uint8_t* command_buffer, uint8_t* response_buffer,
+                                                 size_t length){
+  // Blocking interrupts in order to force a busy wait until the end of the command
+  BlockInterrupts block;
+
+  SetupReception(response_buffer, length);
+  SendCommand(command_buffer, length, true);
+
+  WaitResponse();
+
+  // NOTE - CS has to be high for at least 650ns between bytes.
+  SetChipSelect();
+  SystemTimer::singleton().delay(microseconds(1));
+}
+
+
+// This must only called with interrupts disabled.
+template<size_t MaxSlaves>
+void DaisyChain<MaxSlaves>::WaitResponse(){
+  // Wait until transmission is finished and clear corresponding interrupt flag
+  while (!rx_dma_->InterruptStatus(DMA::Interrupt::TransferComplete)) {
+  }
+
+  // Clear the interrupt flag so I won't get an interrupt as soon as I re-enable them
+  rx_dma_->ClearInterrupt(DMA::Interrupt::Global);
+}
+
+template<size_t MaxSlaves>
+void DaisyChain<MaxSlaves>::on_rx_complete(){
+  // NOTE - CS has to be high for at least 650ns between bytes.
+  // We don't include a busy wait here because in practice, the seting up of the next transmission
+  // takes longer than that.
+  SetChipSelect();
+  ProcessReceivedData();
+  TransmitNextCommand();
+}
+
+template<size_t MaxSlaves>
+void DaisyChain<MaxSlaves>::ProcessReceivedData() {
+  for(size_t slave = 0 ; slave < num_slaves_ ; slave++) {
+    if(current_request_[slave] < queue_count_[slave]) {
+      // We are processing a request for that slave.  Get that request handy.
+      Request *request = &request_queue_[slave][current_request_[slave]];
+
+      // We only record the response if the caller cares about it, and provided proper pointers
+      // where we can put the response and inform him it is valid
+      if(request->response && request->processed && save_response_[slave]) {
+        // Update the response with received data from that slave
+        request->response[response_count_[slave]++] = receive_buffer_[slave];
+      }
+      if(end_of_request_[slave]){
+        // inform the caller that the request is processed
+        if(request->processed) {
```

Comment:
Like here, the "ifs" go several levels deep, and I imagine it would get hard to plan good tests for every branch, or to mentally trace them. Since I don't see any passed parameters that would be of temporary scope but only member variables, maybe it would be safe to break this function apart and  just pass some "slave" index to direct the action?
Also, if there is some set of operations on a common set of buffer containers, maybe it even makes sense to encapsulate all those buffer containers into a yet one more layer that is then member of this class? And then it would be testable separately?

Reply:
I made the logic a bit simpler, the biggest thing I am reusing here is `request` which is indeed derived from members, but making a function out of something like: 
```
if(request->processed) {
  *(request->processed) = true;
}
```
seems overkill here.

Reply:
Yes, some limit to this, but there is nothing wrong with VERY small functions if they conceptually elucidate.